File: service_frontend/types/portal-review.ts
/**
 * Review/Approval SURFACE wire types (plan sprint-4/06 slice 2) — the shapes the
 * My Submissions / My Reviews (two-pane) / Decisions surfaces consume. Mirrors
 * the backend camelCase contracts in `app/schemas/review.py` (MySubmissionOut,
 * MyReviewOut, ReviewTwoPaneOut, DecisionQueueItemOut, …).
 *
 * Identity-agnostic: NOTHING here says "profile" or "user" — the SAME shapes are
 * served to the staff `(protected)` surfaces (user-kind actors) and the portal
 * `(portal)` surfaces (profile-kind actors). The page supplies the right
 * service; the shared components branch on NOTHING (AC-06-46/47/54).
 */
import type { FormDocument, FormAnswers } from '@/types/forms';

/** Submission lifecycle status as resolved by the status engine (label + hex). */
export interface ReviewStatusRef {
  id: string;
  label: string;
  color: string;
}

/** Terminal decision an actor with `can_decide` may take (AC-06-07/45). */
export type ReviewDecision = 'ACCEPTED' | 'REJECTED' | 'REVISIONS';

/** One row in My Submissions (AC-06-43/08) — lifecycle + decision + optional
 * feedback ONLY. Raw reviews/scores are NEVER part of this shape. */
export interface MySubmission {
  groupId: string;
  submissionId: string;
  reviewConfigurationId: string;
  reviewConfigurationName: string;
  /** The config's bound event/context (AC-06-25) — null for a context-less
   * config; `contextLabel` is the real event name for per-row labelling. */
  contextId: string | null;
  contextLabel: string | null;
  formId: string;
  title: string;
  revisionNumber: number;
  statusId: string;
  statusLabel: string;
  statusColor: string;
  decision: ReviewDecision | null;
  feedbackText: string | null;
  decidedAt: string | null;
  canRevise: boolean;
  submittedAt: string | null;
  createdAt: string | null;
}

/** A review configuration the current actor may submit to (AC-06-03) — surfaced
 * even with ZERO prior submissions. `canSubmitNow` reflects the live window +
 * tier-1 block state; the backend is the real gate. */
export type SubmitWindowState = 'open' | 'closed' | 'notYetOpen';

export interface SubmitOption {
  configId: string;
  name: string;
  contextId: string | null;
  contextLabel: string | null;
  submissionFormId: string;
  windowState: SubmitWindowState;
  windowStart: string | null;
  windowEnd: string | null;
  canSubmitNow: boolean;
}

/** The author's view of a submission group's CURRENT revision INCLUDING answers
 * (AC-06-09) — used to prefill the revise form. Lifecycle + decision only; never
 * raw reviews/scores. */
export interface SubmissionDetail extends MySubmission {
  answers: FormAnswers;
  versionId: string;
}

/** Per-config submit entry point shown in My Submissions while a window is open.
 * Derived client-side from the surfaces (one entry per config the actor can
 * submit to); the backend gates the actual submit. */
export interface SubmitFormState {
  state: 'open' | 'closed' | 'blocked' | 'not_eligible';
  configId?: string;
  message?: string;
  form?: {
    formId: string;
    versionId: string;
    name: string;
    description: string | null;
    definition: FormDocument;
    paged: boolean;
  };
}

/** One assignment in the reviewer queue (AC-06-44). `dueAt` = the window end. */
export interface MyReview {
  assignmentId: string;
  reviewConfigurationId: string;
  reviewConfigurationName: string;
  contextId: string | null;
  contextLabel: string | null;
  groupId: string;
  revisionNumber: number;
  status: 'PENDING' | 'COMPLETED' | 'ESCALATED' | 'WITHDRAWN';
  dueAt: string | null;
  assignedAt: string | null;
  completedAt: string | null;
  title: string;
}

/** The read-only submission pane payload (pinned version + answers). */
export interface ReviewSubmissionPane {
  id: string;
  formId: string;
  versionId: string;
  answers: FormAnswers;
  definition: FormDocument | null;
  revisionNumber: number;
}

/** The review-form (rubric) fill pane payload. */
export interface ReviewFormPane {
  formId: string;
  versionId: string;
  versionNumber: number;
  name: string;
  description: string | null;
  definition: FormDocument;
  paged: boolean;
}

/** THIS reviewer's own in-progress draft (reviewer isolation — never another's). */
export interface ReviewDraftPane {
  submissionId: string;
  answers: FormAnswers;
  status?: string;
}

/** The full two-pane grade payload (AC-06-05/48). */
export interface ReviewTwoPane {
  assignmentId: string;
  status: 'PENDING' | 'COMPLETED' | 'ESCALATED' | 'WITHDRAWN';
  reviewConfigurationId: string;
  reviewConfigurationName: string;
  submission: ReviewSubmissionPane;
  reviewForm: ReviewFormPane | null;
  reviewDraft: ReviewDraftPane | null;
}

/** One completed review inside a decision group (decider-only — AC-06-07). */
export interface DecisionReview {
  assignmentId: string;
  reviewSubmissionId: string;
  answers: FormAnswers;
  score: number | null;
  completedAt: string | null;
}

/** One group in the Decisions surface (AC-06-45/07). */
export interface DecisionItem {
  groupId: string;
  submissionId: string;
  reviewConfigurationId: string;
  reviewConfigurationName: string;
  contextId: string | null;
  contextLabel: string | null;
  title: string;
  revisionNumber: number;
  statusId: string;
  statusLabel: string;
  statusColor: string;
  averageScore: number | null;
  completedCount: number;
  requiredReviewCount: number;
  ready: boolean;
  decision: ReviewDecision | null;
  feedbackText: string | null;
  reviews: DecisionReview[];
  submittedAt: string | null;
  createdAt: string | null;
}

/** Save-draft / submit result for a review (assignment). */
export interface ReviewDraftResult {
  submissionId: string;
  answers: FormAnswers;
}

/** Revise result (a fresh Draft revision to fill). */
export interface ReviseResult {
  submissionId: string;
  groupId: string;
  revisionNumber: number;
}

/**
 * The data + actions a surface page passes into a shared presentational
 * component (the two-pane grade view, the decision panel). Identity-agnostic:
 * the page supplies the right service-bound callbacks; the component never knows
 * whether it is staff or portal (AC-06-46/47, reuse-don't-rebuild).
 */
export interface ReviewSurfaceActions {
  /** Resolve the two-pane payload for an assignment. */
  getTwoPane(assignmentId: string): Promise<ReviewTwoPane>;
  /** Save the review draft (resume later). */
  saveDraft(assignmentId: string, answers: FormAnswers): Promise<ReviewDraftResult>;
  /** Submit the review (terminal — assignment → COMPLETED). */
  submitReview(assignmentId: string, answers: FormAnswers): Promise<unknown>;
}

export interface DecisionSurfaceActions {
  /** Take the terminal decision for a group (first-wins; 409 if already decided). */
  decide(
    groupId: string,
    decision: ReviewDecision,
    feedback: string | null,
  ): Promise<DecisionItem>;
}

File: dreamz_ems_backend/modules/ems/routers/portal_review.py
"""Profile-portal review surfaces (plan sprint-4/06 slice 2) — ``/portal/*``.

Profile-authed (``current_profile``) + gated by portal-permission keys
(``require_portal_permission``). HTTP only — Router → ``PortalReviewService`` →
core review/form services. The router NEVER accepts a ``tenant_id`` from the
client (it comes from the Profile JWT, AC-06-14/57).

Wired as a ``"public": true`` manifest router so the loader does NOT inject the
staff ``require_module`` gate — these are Profile-authed, not staff-authed (same
as ``portal``/``portal_auth``).

Surfaces:
- My Submissions (AC-06-43/52/08): list + submit-form view + submit + revise.
- My Reviews (AC-06-44/05/48): queue + two-pane + draft + submit (reviewer-isolated).
- Decisions (AC-06-45/07): queue (live average + ready) + first-wins decide.
"""
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.review_engine.service import (
    ReviewDecisionConflict,
    ReviewTransitionUnavailable,
    ReviewValidationError,
)
from app.schemas.review import (
    DecisionQueueItemOut,
    MyReviewOut,
    MySubmissionOut,
    ReviewDecisionIn,
    ReviewDraftIn,
    ReviewTwoPaneOut,
)
from modules.ems.models import Profile
from modules.ems.portal_deps import require_portal_permission
from modules.ems.portal_review_service import (
    PortalReviewClosed,
    PortalReviewForbidden,
    PortalReviewNotFound,
    PortalReviewService,
)

router = APIRouter()


# ── My Submissions (AC-06-43/52/08) ───────────────────────────────────────────


@router.get("/submissions", response_model=List[MySubmissionOut])
def my_submissions(
    contextId: Optional[str] = None,
    profile: Profile = Depends(require_portal_permission("my_submissions.read")),
    db: Session = Depends(get_db),
) -> List[MySubmissionOut]:
    """My Submissions — narrowed to one event when ``contextId`` is set (AC-06-25),
    else across all contexts (each row carries its event for per-row labelling)."""
    rows = PortalReviewService(db).my_submissions(profile, context_id=contextId)
    return [MySubmissionOut(**r) for r in rows]


@router.get("/submit-options")
def submit_options(
    contextId: Optional[str] = None,
    profile: Profile = Depends(require_portal_permission("forms.submit")),
    db: Session = Depends(get_db),
) -> List[Dict[str, Any]]:
    """Review configurations the current Profile may submit to (AC-06-03) — drives
    the Submit actions on My Submissions even with zero prior submissions.
    ``contextId`` narrows to one event (AC-06-25)."""
    return PortalReviewService(db).submit_options(profile, context_id=contextId)


@router.get("/submissions/{group_id}")
def submission_detail(
    group_id: str,
    profile: Profile = Depends(require_portal_permission("my_submissions.read")),
    db: Session = Depends(get_db),
) -> Dict[str, Any]:
    """The author's own submission group with the current revision's answers
    (AC-06-09 — prefills revise). Author-only; 404 otherwise / cross-tenant."""
    try:
        return PortalReviewService(db).submission_detail(profile, group_id)
    except PortalReviewNotFound:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Submission not found.")


@router.get("/review-configs/{config_id}/submit-form")
def submit_form_view(
    config_id: str,
    profile: Profile = Depends(require_portal_permission("forms.submit")),
    db: Session = Depends(get_db),
) -> Dict[str, Any]:
    try:
        return PortalReviewService(db).submit_form_view(profile, config_id)
    except PortalReviewNotFound:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Review process not found.")


@router.post("/review-configs/{config_id}/submissions", response_model=MySubmissionOut)
def submit(
    config_id: str,
    body: ReviewDraftIn,
    profile: Profile = Depends(require_portal_permission("forms.submit")),
    db: Session = Depends(get_db),
) -> MySubmissionOut:
    from app.services.form_service import FormClosed, FormSubmitInvalid

    try:
        payload = PortalReviewService(db).submit(profile, config_id, body.answers)
    except PortalReviewNotFound:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Review process not found.")
    except PortalReviewForbidden:
        raise HTTPException(status.HTTP_403_FORBIDDEN, "You cannot submit to this process.")
    except PortalReviewClosed as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, str(exc))
    except FormClosed as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, str(exc))
    except FormSubmitInvalid as exc:
        raise HTTPException(status.HTTP_422_UNPROCESSABLE_ENTITY, detail={"fieldErrors": exc.field_errors})
    return MySubmissionOut(**payload)


@router.post("/submissions/{group_id}/revise")
def revise(
    group_id: str,
    profile: Profile = Depends(require_portal_permission("my_submissions.read")),
    db: Session = Depends(get_db),
) -> Dict[str, Any]:
    try:
        return PortalReviewService(db).revise(profile, group_id)
    except PortalReviewNotFound:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Submission not found.")
    except PortalReviewForbidden:
        raise HTTPException(status.HTTP_403_FORBIDDEN, "You cannot revise this submission.")
    except PortalReviewClosed as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, str(exc))


@router.post("/submissions/{submission_id}/resubmit", response_model=MySubmissionOut)
def resubmit(
    submission_id: str,
    body: ReviewDraftIn,
    profile: Profile = Depends(require_portal_permission("my_submissions.read")),
    db: Session = Depends(get_db),
) -> MySubmissionOut:
    from app.services.form_service import FormSubmitInvalid

    try:
        payload = PortalReviewService(db).resubmit(profile, submission_id, body.answers)
    except PortalReviewNotFound:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Submission not found.")
    except PortalReviewForbidden:
        raise HTTPException(status.HTTP_403_FORBIDDEN, "You cannot revise this submission.")
    except PortalReviewClosed as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, str(exc))
    except FormSubmitInvalid as exc:
        raise HTTPException(status.HTTP_422_UNPROCESSABLE_ENTITY, detail={"fieldErrors": exc.field_errors})
    return MySubmissionOut(**payload)


# ── My Reviews (AC-06-44/05/48) ───────────────────────────────────────────────


@router.get("/reviews", response_model=List[MyReviewOut])
def my_reviews(
    contextId: Optional[str] = None,
    profile: Profile = Depends(require_portal_permission("my_reviews.read")),
    db: Session = Depends(get_db),
) -> List[MyReviewOut]:
    rows = PortalReviewService(db).my_reviews(profile, context_id=contextId)
    return [MyReviewOut(**r) for r in rows]


@router.get("/reviews/{assignment_id}", response_model=ReviewTwoPaneOut)
def review_two_pane(
    assignment_id: str,
    profile: Profile = Depends(require_portal_permission("my_reviews.read")),
    db: Session = Depends(get_db),
) -> ReviewTwoPaneOut:
    try:
        payload = PortalReviewService(db).review_two_pane(profile, assignment_id)
    except PortalReviewNotFound:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Review assignment not found.")
    return ReviewTwoPaneOut(**payload)


@router.post("/reviews/{assignment_id}/draft")
def save_review_draft(
    assignment_id: str,
    body: ReviewDraftIn,
    profile: Profile = Depends(require_portal_permission("my_reviews.read")),
    db: Session = Depends(get_db),
) -> Dict[str, Any]:
    try:
        return PortalReviewService(db).save_review_draft(profile, assignment_id, body.answers)
    except PortalReviewNotFound:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Review assignment not found.")
    except PortalReviewClosed as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, str(exc))


@router.post("/reviews/{assignment_id}/submit")
def submit_review(
    assignment_id: str,
    body: ReviewDraftIn,
    profile: Profile = Depends(require_portal_permission("my_reviews.read")),
    db: Session = Depends(get_db),
) -> Dict[str, Any]:
    try:
        return PortalReviewService(db).submit_review(profile, assignment_id, body.answers)
    except PortalReviewNotFound:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Review assignment not found.")
    except PortalReviewClosed as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, str(exc))
    except ReviewValidationError as exc:
        raise HTTPException(status.HTTP_422_UNPROCESSABLE_ENTITY, str(exc))


# ── Decisions (AC-06-45/07) ───────────────────────────────────────────────────


@router.get("/decisions", response_model=List[DecisionQueueItemOut])
def decisions(
    contextId: Optional[str] = None,
    profile: Profile = Depends(require_portal_permission("decisions.read")),
    db: Session = Depends(get_db),
) -> List[DecisionQueueItemOut]:
    rows = PortalReviewService(db).decisions(profile, context_id=contextId)
    return [DecisionQueueItemOut(**r) for r in rows]


@router.post("/decisions/{group_id}/decide", response_model=DecisionQueueItemOut)
def decide(
    group_id: str,
    body: ReviewDecisionIn,
    profile: Profile = Depends(require_portal_permission("decisions.read")),
    db: Session = Depends(get_db),
) -> DecisionQueueItemOut:
    try:
        payload = PortalReviewService(db).decide(profile, group_id, body.decision, body.feedback)
    except PortalReviewNotFound:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Submission group not found.")
    except PortalReviewForbidden:
        raise HTTPException(status.HTTP_403_FORBIDDEN, "You cannot decide on this submission.")
    except ReviewDecisionConflict:
        raise HTTPException(status.HTTP_409_CONFLICT, "A decision has already been made.")
    except ReviewTransitionUnavailable as exc:
        raise HTTPException(status.HTTP_422_UNPROCESSABLE_ENTITY, str(exc))
    except ReviewValidationError as exc:
        raise HTTPException(status.HTTP_422_UNPROCESSABLE_ENTITY, str(exc))
    return DecisionQueueItemOut(**payload)
